pkg/workflow/coordinator/k8sapi/k8sapi.go

```diff
@@ -0,0 +1,132 @@
+package k8sapi
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/caicloud/cyclone/pkg/k8s/clientset"
+	"github.com/caicloud/cyclone/pkg/workflow/coordinator/common"
+)
+
+type K8sapiExector struct {
+	client    clientset.Interface
+	namespace string
+	podName   string
+}
+
+func NewK8sapiExector(n string, pod string, client clientset.Interface) *K8sapiExector {
+	return &K8sapiExector{
+		namespace: n,
+		podName:   pod,
+		client:    client,
+	}
+}
+
+// WaitContainersTerminate waits containers except for 'excepts' to be 'expectState' status.
+func (k *K8sapiExector) WaitContainers(timeout time.Duration, expectState common.ContainerState, excepts []string) error {
+	ticker := time.NewTicker(time.Second * 1)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	log.Infof("Starting to wait for containers of pod %s to be %s ...", k.podName, expectState)
+	for {
+		select {
+		case <-ticker.C:
+			pod, err := k.client.CoreV1().Pods(k.namespace).Get(k.podName, meta_v1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			containerNum := len(pod.Spec.Containers)
+			expectTerminatedNum := containerNum
+			actualNum := 0
+			for _, cs := range pod.Status.ContainerStatuses {
+				for _, except := range excepts {
+					if cs.Name == except {
+						expectTerminatedNum--
+						continue
+					}
+				}
+
+				switch expectState {
+				case common.ContainerStateTerminated:
+					// Check if container is terminated
+					if cs.State.Terminated != nil {
+						log.Infof("Container %s is terminated: %v", cs.Name, cs.State.Terminated)
+						actualNum++
+					}
+				case common.ContainerStateNotWaiting:
+					// Check if container is not waiting
+					if cs.State.Running != nil || cs.State.Terminated != nil {
```

Comment:
Has logic error here?  Should be `if cs.State.Running == nil || cs.State.Terminated == nil`?

Reply:
I do not think so, `cs.State.Running != nil || cs.State.Terminated != nil` indicate it not warting.